Override PHP's default error and exception handler printing

anomiex opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

WordPress relies on PHP's `display_errors` setting for error reporting. One unfortunate result of that is that errors containing HTML may be served to the browser unescaped.

WordPress/WordPress-Coding-Standards tries to help avoid this by having a sniff (WordPress.Security.EscapeOutput) that, among other things, asks that strings passed to `trigger_error()` and `throw new SomeClass()` be escaped. This is not a very good solution for a few reasons:

* It doesn't do anything about errors or exceptions thrown by PHP itself.
* It doesn't do anything about errors or exceptions thrown by libraries that might be bundled into themes or plugins or any custom code blog authors might inject.
* It means that text-based log files, non-HTML emails, CLI output, and the like will contain HTML entities.

Instead we should follow the principle that values should be escaped close to the point of output, when we know what kind of escaping will be needed. Which, in the context of PHP, means using `set_error_handler()` and `set_exception_handler()` to handle the outputting.

But WordPress also relies on various other things that PHP's default handlers do, not all of which we can duplicate. But on the plus side since `display_errors` is not exactly a boolean (and PHP doesn't care anyway) we can have the handlers themselves set `display_errors` to a sentinel value that PHP will interpret as "off", do any necessary output, and then hand off to PHP's default handlers to do everything else.

Trac ticket: https://core.trac.wordpress.org/ticket/59282

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anomiex committed:
Override PHP's default error and exception handler printing

WordPress relies on PHP's `display_errors` setting for error reporting.
One unfortunate result of that is that errors containing HTML may be
served to the browser unescaped.

WordPress/WordPress-Coding-Standards tries to help avoid this by having
a sniff (WordPress.Security.EscapeOutput) that, among other things, asks
that strings passed to `trigger_error()` and `throw new SomeClass()` be
escaped. This is not a very good solution for a few reasons:

* It doesn't do anything about errors or exceptions thrown by PHP
  itself.
* It doesn't do anything about errors or exceptions thrown by libraries
  that might be bundled into themes or plugins or any custom code blog
  authors might inject.
* It means that text-based log files, non-HTML emails, CLI output, and
  the like will contain HTML entities.

Instead we should follow the principle that values should be escaped
close to the point of output, when we know what kind of escaping will be
needed. Which, in the context of PHP, means using `set_error_handler()`
and `set_exception_handler()` to handle the outputting.

But WordPress also relies on various other things that PHP's default
handlers do, not all of which we can duplicate. But on the plus side
since `display_errors` is not exactly a boolean (and PHP doesn't care
anyway) we can have the handlers themselves set `display_errors` to a
sentinel value that PHP will interpret as "off", do any necessary
output, and then hand off to PHP's default handlers to do everything
else.

dingo-d commented on src/wp-error-handling.php:
```diff
@@ -0,0 +1,170 @@
+function wp_error_handler( $errno, $errstr, $errfile, $errline ) {
+	static $display_errors = false;
+
+	$v = ini_get( 'display_errors' );
```
I'd name this variable to something a bit more understandable. Maybe `$display_errors_ini_value`?

anomiex replied:
Meh.

dingo-d replied:
Meh, or not, naming is important, because somebody will have to maintain the code you add to WordPress core. Please be more considerate and think about the code you write. It's read by humans, not by computers.

dingo-d commented on src/wp-error-handling.php:
```diff
@@ -0,0 +1,170 @@
+ *
+ * @param Throwable $ex Object that was thrown.
+ */
+function wp_exception_handler( Throwable $ex ) {
```
Same with the `$ex` variable name here. `$exception` would be better IMO.

anomiex replied:
"$ex" is what PHP's own docs use at https://www.php.net/set_exception_handler#refsect1-function.set-exception-handler-parameters, FYI.

dingo-d reviewed (commented):
Left some minor comments regarding code readability mostly.

I think the new functions should be covered by unit tests.

anomiex committed:
Better match PHP's own logic for interpreting `display_error` strings

And also don't bother replacing falsey values with "wp".

anomiex committed:
Chain to existing handlers, in case someone configured some in PHP's config.

That also requires separating the "print" logic into a separate helper function.

anomiex committed:
Rename some variables per request

anomiex committed:
Fix boolean ini handling

anomiex committed:
Use a namespace, mainly so test can "mock" functions

anomiex committed:
Add tests

paulshryock commented on src/wp-error-handling.php:
```diff
@@ -0,0 +1,243 @@
+ *
+ * Initially we simply want to avoid error messages accidentally containing
+ * HTML in certain configurations. In the future we may improve the display
+ * beyond what PHP does by default.
```
I don't think we need "In the future we may improve the display beyond what PHP does by default."

Rather than speculate about what might be in this class at some point later, we should just describe what is in the class currently.

anomiex replied:
I think I was mainly trying to set expectations here, so people didn't think this shouldn't be improved.

anomiex committed:
Fix werid PHPCS sniff

This seems opposite to how we do anonymous functions, but 🤷.

anomiex committed:
Handle deprecation of E_STRICT in PHP 8.4, while still supporting it for older PHP
